tests/integration/cli/commands/test_issue_2414.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2024 Pex project contributors.
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import os.path
+
+import pytest
+
+from pex.pep_440 import Version
+from pex.pep_503 import ProjectName
+from pex.pip.log_analyzer import ErrorMessage
+from pex.requirements import parse_requirement_string
+from pex.resolve.configured_resolver import ConfiguredResolver
+from pex.resolve.locked_resolve import LockConfiguration, LockStyle
+from pex.resolve.locker import Locker, LockResult
+from pex.resolve.resolved_requirement import (
+    ArtifactURL,
+    Fingerprint,
+    PartialArtifact,
+    Pin,
+    ResolvedRequirement,
+)
+from pex.targets import LocalInterpreter
+from pex.typing import TYPE_CHECKING
+from testing import data
+
+if TYPE_CHECKING:
+    from typing import Any
+
+
+@pytest.fixture
+def locker(tmpdir):
+    # type: (Any) -> Locker
+    download_dir = os.path.join(str(tmpdir), "downloads")
+    return Locker(
+        target=LocalInterpreter.create(),
+        root_requirements=[parse_requirement_string("wheel")],
+        resolver=ConfiguredResolver.default(),
+        lock_configuration=LockConfiguration(style=LockStyle.SOURCES),
+        download_dir=download_dir,
+    )
+
+
+def analyze_log(
+    locker,  # type: Locker
+    log_name,  # type: str
+):
+    # type: (...) -> None
+    with open(data.path("pip_logs", log_name)) as fp:
+        for line in fp:
+            result = locker.analyze(line)
+            assert not isinstance(result.data, ErrorMessage)
+    locker.analysis_completed()
+
+
+@pytest.mark.parametrize("pip_version", ["23.2", "23.3.1"])
```

Comment:
Pre fix commits this ran like:
```
:; tox -epy311-integration -- --devpi -vvsk test_issue_2414 -n0
...
tests/integration/cli/commands/test_issue_2414.py::test_redirects_dont_stomp_original_index_urls[23.2] PASSED
tests/integration/cli/commands/test_issue_2414.py::test_redirects_dont_stomp_original_index_urls[23.3.1] FAILED
...
```

Which is expected as per the OP in #2414. The failure for the 23.3.1 case looked like:
```
...
           Full diff:
E           - (ResolvedRequirement(pin=Pin(project_name=ProjectName(raw='wheel', validated=False, normalized='wheel'), version=Version(raw='0.43.0', normalized='0.43')), artifact=PartialArtifact(url=ArtifactURL(raw_url='https://files.pythonhosted.org/packages/7d/cd/d7460c9a869b16c3dd4e1e403cce337df165368c71d6af229a74699622ce/wheel-0.43.0-py3-none-any.whl', download_url='https://files.pythonhosted.org/packages/7d/cd/d7460c9a869b16c3dd4e1e403cce337df165368c71d6af229a74699622ce/wheel-0.43.0-py3-none-any.whl', normalized_url='https://files.pythonhosted.org/packages/7d/cd/d7460c9a869b16c3dd4e1e403cce337df165368c71d6af229a74699622ce/wheel-0.43.0-py3-none-any.whl', scheme='https', path='/packages/7d/cd/d7460c9a869b16c3dd4e1e403cce337df165368c71d6af229a74699622ce/wheel-0.43.0-py3-none-any.whl', fragment_parameters={}, fingerprints=()), fingerprint=None, verified=False), additional_artifacts=(PartialArtifact(url=ArtifactURL(raw_url='https://m.devpi.net/root/pypi/%2Bf/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', download_url='https://m.devpi.net/root/pypi/%2Bf/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', normalized_url='https://m.devpi.net/root/pypi/+f/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', scheme='https', path='/root/pypi/+f/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', fragment_parameters={}, fingerprints=()), fingerprint=Fingerprint(algorithm='sha256', hash='465ef92c69fa5c5da2d1cf8ac40559a8c940886afcef87dcf14b9470862f1d85'), verified=False),)),)
E           + (ResolvedRequirement(pin=Pin(project_name=ProjectName(raw='wheel', validated=False, normalized='wheel'), version=Version(raw='0.43.0', normalized='0.43')), artifact=PartialArtifact(url=ArtifactURL(raw_url='https://m.devpi.net/root/pypi/%2Bf/55c/570405f142630/wheel-0.43.0-py3-none-any.whl', download_url='https://m.devpi.net/root/pypi/%2Bf/55c/570405f142630/wheel-0.43.0-py3-none-any.whl', normalized_url='https://m.devpi.net/root/pypi/+f/55c/570405f142630/wheel-0.43.0-py3-none-any.whl', scheme='https', path='/root/pypi/+f/55c/570405f142630/wheel-0.43.0-py3-none-any.whl', fragment_parameters={}, fingerprints=()), fingerprint=Fingerprint(algorithm='sha256', hash='55c570405f142630c6b9f72fe09d9b67cf1477fcf543ae5b8dcb1f5b7377da81'), verified=False), additional_artifacts=(PartialArtifact(url=ArtifactURL(raw_url='https://m.devpi.net/root/pypi/%2Bf/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', download_url='https://m.devpi.net/root/pypi/%2Bf/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', normalized_url='https://m.devpi.net/root/pypi/+f/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', scheme='https', path='/root/pypi/+f/465/ef92c69fa5c5d/wheel-0.43.0.tar.gz', fragment_parameters={}, fingerprints=()), fingerprint=Fingerprint(algorithm='sha256', hash='465ef92c69fa5c5da2d1cf8ac40559a8c940886afcef87dcf14b9470862f1d85'), verified=False),)),)

tests/integration/cli/commands/test_issue_2414.py:93: AssertionError
...
```

Note the whl URLs differed as in the OP in #2414 case with re-directed https://files.pythonhosted.org URLs vs the expected https://m.devpi.net URLs.